### PR TITLE
Add support for CASE WHEN ... ELSE ... END expression

### PR DESCRIPTION
**Pull request by Ten0**
This is part of the SQL standard so the syntax is the same for all backends.
[PostgreSQL](https://www.postgresql.org/docs/current/functions-conditional.html#FUNCTIONS-CASE)
[MySQL](https://mariadb.com/kb/en/case-operator/)
[SQLite](https://www.sqlite.org/lang_expr.html#the_case_expression)

I'm not super happy with the amount of duplicated code compared to other operators, but because of how unusual the SQL writing is:
```rust
out.push_sql("CASE WHEN ");
self.condition.walk_ast(out.reborrow())?;
out.push_sql(" THEN ");
self.if_true.walk_ast(out.reborrow())?;
out.push_sql(" ELSE ");
self.if_false.walk_ast(out.reborrow())?;
out.push_sql(" END");
```

it didn't quite fit in the existing operators macros.

It may be possible to clear this duplication by creating derives for `SelectableExpression`/`AppearsOnTable`/`FieldAliasMapper`.

I have opted to keep it simple and not implemented the auto nullability propagation (which would be hard to implement with regards to its interactions with auto_type, introducing the same complex issues as we had with And/Or) so it's left up to the user to make the nullability of the `if_true` and the `if_false` match using `.nullable()` if necessary.

### CODE REVIEW

**Comment by weiznich** [comment]
I think this is a reasonable addition to diesel as this is a pretty common SQL construct. I'm not sure if we want to generalize it a bit further to allow expressing the following variants as well:

* Skipping the `ELSE` branch and therefore make the expression `Nullable`
* Adding additional `WHEN condition THEN value` branches, e.g via a builder pattern, represented by nested type

Unrelated question: Are you at EuroRust? If yes we could try to meet there tomorrow

**Comment by Ten0** [comment]
> * Skipping the `ELSE` branch and therefore make the expression `Nullable`

What I had in mind for this extension was to name the associated function just `case_when`, hence the name of the module and of `case_when_else`, preparing a spot for this future extension.

> Adding additional WHEN condition THEN value branches, e.g via a builder pattern, represented by nested type

Hmm I didn't have an appropriate design come to mind initially but I it looks like this would work.
Thinking about it some more, I think I like the following design even more:
```rust
case_when_else(((cond_a, if_a), (cond_b, if_b)), otherwise)
```
which would avoid having nesting. With the builder pattern since the type has to change it doesn't bring `fold` capability anyway so the flat writing may be just simpler.

It would imply updating the current syntax to:
```rust
case_when_else((condition, if_true), otherwise)
```

> Unrelated question: Are you at EuroRust? If yes we could try to meet there tomorrow

Oh! Would have been a pleasure but unfortunately I'm not there. 🙁 Maybe next year! Wasn't too inspired by last year's talks but this year's program looks amazing!

**Comment by Ten0** [comment]
So I went ahead and implemented it.

The tuple version didn't work too well due to conflicting impls considerations and required macros with implementations on tuples so I eventually went with the recursive type approach.

**Review by weiznich — APPROVED** [review]
Looks good beside the minor comment about the `else` method name. Thanks for finishing this up.

File: diesel/src/expression/case_when.rs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+            else_expr: self.else_expr,
+        }
+    }
+}
+
+impl<Whens> CaseWhen<Whens, NoElseExpression> {
+    /// Sets the `ELSE` branch of the `CASE` expression
+    ///
+    /// See the [`case_when`] documentation for more details.
+    pub fn else_<E>(self, if_no_other_branch_matched: E) -> helper_types::Else_<Self, E>
```
[the remark this answers is not in the document]
Comment (Ten0) [reply diesel/src/expression/case_when.rs]: 6f45bae8a94487489fc374a81cd2a3a6a5c14e84

**Pull request merged**